File: world.js
const { IncomingMessage } = require("http");

class World {
    constructor(data) {
        this.id_to_items = {};
        data.forEach((value, i) => {
            value['id'] = i + 1,
            this.id_to_items[i+1] = value; 
        })
        this.print = console.log;
    }

    all_items() {
        let lst = [];
        Object.entries(this.id_to_items).forEach((pair) => {
            lst.push(pair[1]);
        })
        return lst;
    }

    item(item_id) {
        var found = null;
        this.all_items().forEach(item => {
            if (item.id == item_id) found=item;
        });
        return found;
    } 

    has_name(item, name) {
        if (!item.id) {
            item = world.item(item);
        }
        return item.name.toLowerCase() === name.toLowerCase();
    }

    item_by_name(name, items) {
        if (!items) items = this.all_items();
        items = items.filter(item => this.has_name(item, name));
        if (items.length > 0) {
            return items[0];
        }
        return null;
    }

    items_in(container) {
        const container_id = container.id || container;
        let lst = []
        this.all_items().forEach(item => {
            if (item["container_id"] === container_id) {
                lst.push(item);
            }
        })
        return lst;
    }

    visible_items_in(container) {
        let items = []
        this.items_in(container).forEach(item => {
            items.push(item);
            if (item.closed) return;
            items = items.concat(this.visible_items_in(item));
        });
        return items;
    }

    is_reachable_by(item, player) {
        if (!item.id) {
            item = this.item(item);
        }
        if (item.container_id === player.id || item.container_id === player.container_id) {
            return true;
        }
        if (!item.container_id) return false;
        const container = this.item(item.container_id);
        if (container.closed) return false;
        return this.is_reachable_by(container, player);
    }

    look_at(player, item) {
        const print = this.print;
        print(item.description || item.name);
        if (item.closed) {
            if (item.locked) {
                print("è chiuso a chiave");
            } else {
                print("è chiuso");
            }
            return;
        }
        let objs = this.items_in(item).filter(obj => (obj.id != player.id));
        if (objs.length === 0) {
            if (item.attributes.includes("container"))  {
                if (item.attributes.includes("living")) {
                    print("è a mani vuote");
                } else {
                    print("...è vuoto");
                }
            } 
        } else if (objs.length === 1) {
            if (item.attributes.includes("living")) {
                print("ha " + objs[0].name);
            } else {
                print("c'è " + objs[0].name);
            }
        } else {
            if (item.attributes.includes("living")) {
                print("ha");
            } else {
                print("ci sono");
            }
            objs.forEach(obj => {
                print("- " + obj.name);
            })
        }
    }
}

exports.World = World;

File: world.py
class Item(dict):
    def __init__(self, world, id, d):
        super().__init__(d)
        self.id = id
        self.world = world

    def __getitem__(self, key):
        try:
            return super().__getitem__(key)
        except KeyError:
            try:
                return self.world.item(super().__getitem__(key + '_id'))
            except KeyError:
                self[key] = None 
                return None

    def __getattr__(self, key):
        return self[key]

    def items(self):
        return [obj for obj in self.world.items() if obj.container == self]

    def visible_items(self):
        l = []
        for o in self.items():
            l.append(o)
            if not o.closed:
                l = l + o.visible_items()
        return l

    def is_reachable_by(self, player):
        if self.container == player or self.container == player.container:
            return True
        elif self.container.is_reachable_by(player) and not self.container.closed:
            return True
        else:
            return False 

    def set_passage(self, room, dir, rev_dir=None):
        if rev_dir is None:
            rev_dir = {'n':'s', 's':'n', 'e':'w', 'w':'e', 'u':'d', 'd':'u'}[dir]
        if self.passages is None:
            self['passages'] = {}
        self.passages[dir] = room.id
        if rev_dir is not False:
            room.set_passage(self, rev_dir, False)

    def be(self, attribute): 
        return self.attributes and (attribute in self.attributes)            


class World:
    def __init__(self):
        self.MAX_ID = 0
        self.id_to_items = {}

    def items(self):
        return self.id_to_items.values()

    def item(self, id):
        return self.id_to_items[id]
    
    def add(self, item):
        self.MAX_ID += 1
        if not isinstance(item, Item):
            item = Item(self, self.MAX_ID, item)
        self.id_to_items[self.MAX_ID] = item
        return item

    def look_at(self, player, item):
        print(item.description or item.name)
        if item.closed:
            if item.locked:
                print("è chiuso a chiave")
            else:
                print("è chiuso")
            return
        objs = [x for x in item.items() if x != player]
        if len(objs) == 0:
            if "container" in item.attributes:
                if "living" in item.attributes:
                    print("è a mani vuote")
                else:
                    print("...è vuoto")
        elif len(objs) == 1:
            if "living" in item.attributes:
                print("ha", objs[0].name)
            else:
                print("c'è", objs[0].name)
        else:
            if "living" in item.attributes:
                print("ha", objs[0].name)
            else:
                print("ci sono:")
            for obj in objs:
                print("-", obj.name)
